BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,175 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/S3SaveRestoreLib.h>
+#include <Library/IoLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Library/BoardInitLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Base.h>
+
+/**
+  This function uploads information in TSEG area
+  designated for S3 save/restore purpose.
+
+  @param    DataPtr       Address of the structure to be copied to TSEG
+
+  @retval   None
+**/
+VOID
+EFIAPI
+SetS3Info (
+  IN  VOID    *DataPtr
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    if ( ((BL_PLD_COMM_HDR *)Ptr)->Signature == BL_PLD_COMM_SIG) {
+      Ptr += ((BL_PLD_COMM_HDR *)Ptr)->TotalSize;
+      continue;
+    }
+    break;
+  }
+
+  if ( Ptr + ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize > SmmBase + SmmSize ) {
+    DEBUG ((DEBUG_ERROR, "Overflowing SMM region\n"));
+    return;
+  }
+  CopyMem ((VOID *)Ptr, DataPtr, ((BL_PLD_COMM_HDR *)DataPtr)->TotalSize);
+}
+
+
+/**
+  Get pointer to a specific struct area in TSEG
+
+  @param      Id                Id of the struct to be retreived
+
+  @retval     NULL              Struct with signature not found
+  @retval     Ptr               Pointer to struct with signature
+**/
+VOID *
+EFIAPI
+GetS3Info (
+  IN  UINT8   Id
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+  BOOLEAN           FoundInfo = FALSE;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return NULL;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
```

Comment:
There is no reason to search the whole SmmSize region.  If as soon as BL_PLD_COMM_SIG is not seen,  the code can break out.